### ISSUE TITLE
missing ability to batch storage requests

### ISSUE THREAD

**Issue by josepot** [paritytech/json-rpc-interface-spec#48]
When building dApps it usually happens that we must query several storage entries at once.

For instance, let's imagine a situation where the dApp first queries the storage entry of `Session.Validators` in order to query the current set of validators.

Once we have the current validators, we may want to do things like:
- accessing the storage values of `Staking.Validators` for all current validators.
- accessing the storage values of `Identity.IdentityOf` for all the current validators.
- etc

Currently, the JSON-RPC client would have to send a `chainHead_unstable_storage` request per each storage entry. Which is unpractical, it generates a ton of unnecessary overhead and it also slows down by an order of magnitude the time that it takes to get the responses. For instance, currently smoldot "chokes" when the client tries to send so many requests at once, forcing the client to do all sorts of shenanigans in order to get all those results.

Ideally, it should be possible for the client to send many requests in one go, and also [receive the responses in batches](https://github.com/paritytech/json-rpc-interface-spec/issues/47).

I'm not sure what the best API for this should be, though. I have a couple of ideas:

- One idea is to add a valid type `values` into `type` argument and a new `subKeys` argument, for querying a bunch of values from the same dictionary. Then in the `key` the user would pass the key of the dictionary (e.g: for accessing `Staking.Validators` the key would be `0x5f3e4907f716ac89b6347d15ececedca88dcde934c658227ee1dfafcd6e16903`) and then in the individual keys of the dictionary (without prepending the key of the dictionary). For Example:

Querying `Staking.Validators` for the following validators: `16iD4v1FUHMauFB17MsrEFGMrEyGtWEwvgyZ6jCLC5eJdEiZ`, `16ky64Ktu3NWtjrVUZV2RDVe46GvhnUGrxqFrYAQZeB7LHHR` and `16mM7YxjyYqXu7uaJfmfCk9xcr8qqaYUyLazTBt3bueQJAFR` would look like this:

```json
{
   "followSubscription":  0,
   "hash": "0xd484270399fd0ac2e0059704ef231e4da08e038602f71590fd21849ad9ef14fd",
   "type": "values",
   "key": "0x5f3e4907f716ac89b6347d15ececedca88dcde934c658227ee1dfafcd6e16903",
   "subKeys": [
     "0x735e16aa65066f1ffc9ada02baa2ef7a2a24a91f4fdadbbe83a22733dd3ae2deda97f45b4c30dc67",
     "0xa0b1a18158a17b4bfeb586a6d2406aeb67d8dc3895f09dde186e89556ec3c166b6fb1d94bd7a025e",
     "0x19eb0157d3d3179ffeffabd8f3747b5ca19766866bd59856a2e65c537c6456c6f6863c026915f473"
   ]
}
```

- Another option (and probably simplest) would be to allow passing an array of `keys` rather than just one `key` when the type is `values`.

Both options would work perfectly well for my needs, TBH.

I'm a bit worried about how convoluted `chainHead_unstable_storage` can end up becoming with so many overloads and arguments/responses permutations. On one hand it kinda feels like it should be broken down into different (and more specific) functions, on the other hand it kinda makes sense to group all the storage functionality under the same umbrella...

I really don't have any strong opinions about the shape of the final API. The only things that I feel strongly about are:

- It should be possible to send a batch of keys to query many storage entries in one request.
- The addition of the [`values` notification](https://github.com/paritytech/json-rpc-interface-spec/issues/47) (or a notification that serves the same purpose).

**Comment by tomaka** [comment]
> it also slows down by an order of magnitude the time that it takes to get the responses

To comment on that: note that smoldot can't fulfill a huge number of requests in parallel anyway, due to limits on the libp2p networking level, which themselves exist to prevent light clients from overloading full nodes.

In other words, sending for example 5000 requests at once might not take less time than sending 100 requests, waiting for the responses, sending 100 requests, waiting for the responses, etc.

> For instance, currently smoldot "chokes" when the client tries to send so many requests at once, forcing the client to do all sorts of shenanigans in order to get all those results.

This is no longer the case in the latest version of smoldot.

> Another option (and probably simplest) would be to allow passing an array of keys rather than just one key when the type is values.

I would be in favor of this option, due to its simplicity. I don't see why how this relates to the `type` being `values`. You can do that for all types of requests.

**Comment by tomaka** [comment]
The documentation of `chainHead_unstable_storage` currently says:

> For optimization purposes, the JSON-RPC server is allowed to wait a little bit (e.g. up to 100ms) before starting to try fulfill the storage request, in order to batch multiple storage requests together.

Because of this, I also opted for sending storage requests one by one.

The reason for this batching is that the JSON-RPC server should be able to group together multiple requests from multiple JSON-RPC clients.
However, a full node JSON-RPC server simply reads from its database, in which case batching is unnecessary, while a light client JSON-RPC server is normally only used by one client at a time.

In other words, I think we should remove this automatic batching and instead indeed allow the client to request multiple keys at the same time.

**Comment by josepot** [comment]
> I would be in favor of this option, due to its simplicity. However I don't see why how this relates to the type being values. You can do that for all types of requests.

Fair enough! Yes, that makes sense. So, just to make sure that we are on the same page, what you are suggesting is to leave the parameter `type` as it is, and then allow the parameter `key` to take either an Array of keys or just a key. Would it make sense to rename the parameter to `keys` and force it to always take a non-empty Array? or should the name of the parameter be renamed? or just leave it as it is and allow either a key or a non-empty array of keys? :thinking:

> In other words, I think we should remove this automatic batching and instead indeed allow the client to request multiple keys at the same time.

I couldn't agree more! The JSON-RPC client can provide abstractions that provide a better understanding about how the data is being consumed, and thus is better suited for batching the requests optimally.
